firmware/library/L3_HAL/sddriver.hpp

```diff
@@ -0,0 +1,239 @@
+        kDelTo      = 0x40 | 33,    // CMD33: set address of the end block for
+                                    // deletion
+        kDel        = 0x40 | 38,    // CMD38: begin deletion from the block
+                                    // range specified by the vector
+                                    // [DEL_FROM : DEL_TO]
+        kAcBegin    = 0x40 | 55,    // CMD55: signals the start of an
+                                    // application-specific command
+        kGetOcr     = 0x40 | 58,    // CMD58: request data from the
+                                    // operational conditions register
+        kAcInit     = 0x40 | 41     // ACMD41: application-specific version of
+                                    // CMD1 (must precede with CMD55)
+    };
+
+    // @function        SendCmd()
+    // @description     sends a command frame to the SD Card and sets the
+    //                  response
+    // @parameter       (SdCommand) sdc                 the command code to
+    //                                                  send as the first byte
+    //                  (uint32_t) arg                  the 4-byte argument
+    //                                                  for the command
+    //                  ~(uint8_t*) response_buffer     pointer to a buffer to
+    //                                                  house the response
+    //                                                  frame
+    //                  ~(uint32_t) delay               delay (ms) between
+    //                                                  asserting CS and
+    //                                                  sending a cmd frame
+    //                  ~(bool) keep_alive              determines whether to
+    //                                                  de-assert CS after
+    //                                                  receiving response
+    // @returns         on success: the number of bytes given in the response
+    //                  on failure: -1
+    virtual uint32_t SendCmd(
+        SdCommand sdc,
+        uint32_t arg,
+        uint8_t response_buffer[] = nullptr,
+        uint32_t delay = 0,
+        bool keep_alive = false) = 0;
+
+    // @function        InitializeHardware()
+    // @description     initializes all required hardware constructs to talk
+    //                  to the SD Card via SPI
+    // @parameter       (?)
+    // @returns         (?)
+    virtual void InitializeHardware() = 0;
+
+    // @function        InitializeSdCard()
+    // @description     runs the SD Card initialization sequence, effectively
+    //                  resetting communication with the card
+    // @parameter       (SdCard*) sd                    a pointer to the SD
+    //                                                  Card data structure
+    // @returns         on init success: true
+    //                  on init failure: false
+    virtual bool InitializeSdCard(SdCard_t* sd) = 0;
```

Comment:
Why do you need to pass an SdCard interface? Shouldn't this simply initialize itself?